overlord/configstate/configcore/debug.go

```diff
@@ -0,0 +1,166 @@
+	supportedConfigurations[coreOptionDebugSystemdLogLevel] = true
+}
+
+func optionChanged(tr RunTransaction, fullOptionName string) bool {
+	for _, name := range tr.Changes() {
+		if name == fullOptionName {
+			return true
+		}
+	}
+	return false
+}
+
+func validateDebugSnapdLogSetting(tr RunTransaction) error {
+	return validateBoolFlag(tr, optionDebugSnapdLog)
+}
+
+func handleDebugSnapdLogConfiguration(tr RunTransaction, opts *fsOnlyContext) error {
+	// Run only if the option changed to avoid extra filesystem access
+	if !optionChanged(tr, coreOptionDebugSnapdLog) {
+		return nil
+	}
+
+	debugLog, err := coreCfg(tr, optionDebugSnapdLog)
+	if err != nil {
+		return err
+	}
+
+	rootDir := dirs.GlobalRootDir
+	if opts != nil {
+		rootDir = opts.RootDir
+	}
+	envDir := filepath.Join(dirs.SnapdStateDir(rootDir), "environment")
+
+	snapdEnvPath := filepath.Join(envDir, "snapd.conf")
+
+	switch debugLog {
+	case "true":
+		if err := os.Mkdir(envDir, 0755); err != nil && !os.IsExist(err) {
+			return err
+		}
+		if err := osutil.EnsureFileState(snapdEnvPath, &osutil.MemoryFileState{
+			Content: []byte("SNAPD_DEBUG=1\n"),
+			Mode:    os.FileMode(0644),
+		}); err != nil {
+			return err
+		}
+	case "false", "":
+		// We simply remove the env file as for the moment we use it
+		// just for SNAPD_DEBUG. If we change that we will need to
+		// locate the variable in the file and remove just that.
+		if err := os.Remove(snapdEnvPath); err != nil && !os.IsNotExist(err) {
+			return err
+		}
+	default:
+		return fmt.Errorf("%s must be true of false, not: %q", optionDebugSnapdLog, debugLog)
+	}
+
+	// TODO change logger to debug enabled. Change
+	// tests/core/debug/task.yaml too when this is done.
+
+	return nil
+}
+
+func validateDebugSystemdLogLevelSetting(tr RunTransaction) error {
+	value, err := coreCfg(tr, optionDebugSystemdLogLevel)
+	if err != nil {
+		return err
+	}
+
+	switch value {
+	case "emerg", "alert", "crit", "err", "warning", "notice", "info", "debug",
+		"0", "1", "2", "3", "45", "6", "7", "":
+		// noop
+	default:
+		return fmt.Errorf("%q is not a valid value for %s (see systemd(1))",
+			value, optionDebugSystemdLogLevel)
+	}
+	return nil
+}
+
+func handleDebugSystemdLogLevelConfiguration(tr RunTransaction, opts *fsOnlyContext) error {
+	// Run only if the option changed to avoid extra filesystem
+	// access / systemctl calls
+	if !optionChanged(tr, coreOptionDebugSystemdLogLevel) {
+		return nil
+	}
+
+	logLevel, err := coreCfg(tr, optionDebugSystemdLogLevel)
+	if err != nil {
+		return err
+	}
+
+	var sysd systemd.Systemd
+	confDir := ""
+	if opts != nil {
+		confDir = dirs.SnapSystemdConfDirUnder(opts.RootDir)
+		sysd = systemd.NewEmulationMode(opts.RootDir)
+	} else {
+		confDir = dirs.SnapSystemdConfDir
+		sysd = systemd.New(systemd.SystemMode, &sysdLogger{})
+	}
+	confFile := filepath.Join(confDir, "20-debug_systemd_log-level.conf")
+
+	if logLevel == "" {
+		// On unsetting, remove the file.
+		os.Remove(confFile)
```

Comment:
Should this return or log the error if it's not because it's missing?

Reply:
I've added a trace in case this happens, I think that makes sense